api_log/models/api_log.py

```diff
@@ -0,0 +1,274 @@
+                collection_id = False
+            log.collection_model = collection_model
+            log.collection_id = collection_id
+
+    @api.model
+    def _headers_hidden_keys(self):
+        """Header keys that should not be logged.
+
+        They might contains sensitive data.
+        """
+        return (
+            "Api-Key",
+            "Cookie",
+        )
+
+    @api.model
+    def _sanitize_headers_dict(self, headers_dict):
+        keys_to_hide = self._headers_hidden_keys()
+        for key in headers_dict:
+            if key in keys_to_hide:
+                headers_dict[key] = "<redacted>"
+        return headers_dict
+
+    @api.model
+    def _headers_to_dict(self, headers):
+        headers_dict = {key: value for key, value in headers.items()}
+        return self._sanitize_headers_dict(headers_dict)
+
+    def _current_time(self):
+        return time.time_ns() / 1e9
+
+    @api.model
+    def _get_http_request(self, request):
+        return request.httprequest
+
+    @api.model
+    def _get_request_body(self, request):
+        """Take extra care with the request's body because it might get consumed."""
+        httprequest = self._get_http_request(request)
+        return httprequest.data
+
+    @api.model
+    def _prepare_log_request(self, request):
+        httprequest = self._get_http_request(request)
+        log_request_values = {
+            "request_url": httprequest.url,
+            "request_method": httprequest.method,
+            "request_headers": self._headers_to_dict(httprequest.headers),
+            "request_body": self._get_request_body(request),
+            "request_date": fields.Datetime.now(),
+            "request_time": self._current_time(),
+        }
+        return log_request_values
+
+    @api.model
+    def log_request(self, request, override_log_values=None):
+        log_request_values = self._prepare_log_request(request)
+        log_request_values.update(override_log_values or {})
+        return self.sudo().create(log_request_values)
+
+    def _inject_log_entry(self, values_dict):
+        values_dict["API-Log-Entry-ID"] = str(self.id)
+        return values_dict
+
+    def _prepare_log_response(self, response):
+        self._inject_log_entry(response.headers)
+        headers_dict = self._headers_to_dict(response.headers)
+        return {
+            "response_status_code": response.status_code,
+            "response_headers": headers_dict,
+            "response_body": response.data,
+            "response_date": fields.Datetime.now(),
+            "response_time": self._current_time(),
+        }
+
+    def log_response(self, response):
+        log_response_values = self._prepare_log_response(response)
+        return self.sudo().write(log_response_values)
+
+    def _prepare_log_exception(self, exception):
+        values = {
+            "stack_trace": "".join(format_exception(exception)),
+            "response_headers": self._inject_log_entry(dict()),
```

Comment:
I think it is especially important to return the header in case of exception, luckily we can do this by setting exception headers attribute, see: https://github.com/OCA/rest-framework/blob/18.0/fastapi/fastapi_dispatcher.py#L46

```suggestion
        exception.headers = getattr(exception, "headers", {})
        values = {
            "stack_trace": "".join(format_exception(exception)),
            "response_headers": self._inject_log_entry(exception.headers),
        }
```
```

Reply:
I agree, that's why I mentioned it specifically, but I didn't know a way to inject it; the solution you proposed is awesome and works great!
I included it, please check the new changes.